### ISSUE TITLE
Support XMP sidecar files

### ISSUE THREAD

**Issue by properlypurple** [bpatrik/pigallery2#156]
**Is your feature request related to a problem? Please describe.**
I use Digikam and Darktable both to manage my photos, and I like keeping the metadata in XMP sidecar files, which makes it a lot easier to maintain interoperability, as well as file syncing and versioning for backup. Currently Pigallery2 reads tags directly from the image file, but ignores XMP sidecar files, thus making it not very useful for tags/faces, etc

**Describe the solution you'd like**
When the metadata is loaded, look for and load from xmp sidecar files as well as the image.

Pigallery2 is the best tool I have found yet, to maintain a read-only gallery for my already managed filesystem hierarchy of photos. I already have a syncing and backup system in place, as well as the apps to manage the photos. Support for sidecar files will make this feature complete for me. I'm happy to provide sample images and files for development/testing.

**Comment by bpatrik** [comment]
This is actually a great idea. 

Would you be interested contributing this to the gallery? I can give you some pointers, shouldn't be that complicated. 
If not, I'll live this issue open and once I catch up with my life, I'll implement it.

**Comment by properlypurple** [comment]
Thanks @bpatrik . My javascript is not great, but I'd love to give this a try. I'll set up a dev environment some time this week and will look into this. I'd love any pointers you can give me.

**Comment by bpatrik** [comment]
Hi,
Sorry for the late reply. If you still have the time to do this feature here are the pointers:

dev env:
1) download source
2) `npm install` to get all dependencies
3) `npm run run-dev`  (you can kill the process once it is finished building the fronted as you not going to change that)
4) `npm run build-backend` this only compiles the typescript files, you can also run `tsc -w` if zou have typescript installed globally
4) `npm start` to run the server

Where to change:
The metadata loading is happening in the MetadataLoader.ts file: https://github.com/bpatrik/pigallery2/blob/af012060ec420fabd8938797a38b6210dc40d3a4/src/backend/model/threading/MetadataLoader.ts#L80
I'm reading the beginning of the photo file here: https://github.com/bpatrik/pigallery2/blob/af012060ec420fabd8938797a38b6210dc40d3a4/src/backend/model/threading/MetadataLoader.ts#L84-L85
Then I'm using multiple libraries to parse the metadata (unfortunately there is not a good what that can do everything on its own). I'm using these libs: `ts-exif-parser, ts-node-iptc, exifreader, image-size`.
Please check if any of those support reading the sidecar files, before adding a new dependency to the project. 

If you update the `metadata` variable in the `loadPhotoMetadata(fullPath: string)` function with the data from the sidecar files, than the site will handle it properly, you do not need to do any other changes.

**Comment by properlypurple** [comment]
I did some tinkering, and it looks like none of the existing dependencies support reading xmp files. I was able to find this
https://github.com/PhilipF5/xmp-sidecar

but I wasn't able to get it working properly. I'm going to start taking out time to try and work for this, but I have no high hopes. Thank you for the detailed instructions(maybe they should go in the readme, or a wiki page.)

**Comment by mcdamo** [comment]
I'm interested in supporting XMP sidecars because this will allow metadata for **videos** such as tags and descriptions.

[xmp-sidecar](https://github.com/PhilipF5/xmp-sidecar) doesn't appear to decode much of the metadata so you would have to manually comb through the raw xml structures anyway.

[exiftool](https://exiftool.org/) is probably a better option, it reads and decodes all file types including XMP sidecars and is interfaced by packages such as [node-exiftool](https://www.npmjs.com/package/node-exiftool) and [exiftool-vendored](https://www.npmjs.com/package/exiftool-vendored).

**Comment by mattiasw** [comment]
Do you have an example XMP sidecar file that I can look at for ExifReader? I'm wondering how complicated it would be to add that functionality.

**Comment by mcdamo** [comment]
> Do you have an example XMP sidecar file that I can look at for ExifReader? I'm wondering how complicated it would be to add that functionality.

XMP is a [standard](https://en.wikipedia.org/wiki/Extensible_Metadata_Platform), you can [generate from exiftool](https://exiftool.org/metafiles.html#xmp) or find [samples](https://github.com/trimmer-io/go-xmp/tree/master/samples) all over the web.

**Comment by bpatrik** [comment]
@mattiasw there is also a similar issue at exifr github with some example: https://github.com/MikeKovarik/exifr/issues/62

**Comment by kagahd** [comment]
Since [exifr#62](https://github.com/MikeKovarik/exifr/issues/62#issuecomment-869568888) is able to read sidecar files, will reading XMP sidecar files be supported by PiGallery2 in the near future? 
That would be awesome!

**Comment by bpatrik** [comment]
Its on my (unwritten) roadmap, but I cant give a deadline.

**Comment by Ultra2D** [comment]
Just started looking into hosting some of my photos online. So far I am really digging Pigallery2. My workflow would be Digicam for face detection and geolocation, then Darktable for further editing. I'd like to keep the original files as is, and use sidecars.

> If you update the `metadata` variable in the `loadPhotoMetadata(fullPath: string)` function with the data from the sidecar files, than the site will handle it properly, you do not need to do any other changes.

Are you saying that _if_ all metadata is contained in the sidecar XMP, I could 'just' try to load from the sidecar instead of the photo? Hacking something like that in as a test is about my level of 'coding' .

**Comment by p2kmgcl** [comment]
Anyone working on this? I'd be happy to contribute to this feature

**Comment by Wihlelm** [comment]
I was also expecting an update on this ticket for many months, and a few days ago, I tried the docker image pigallery2:edge-alpine. As far as I can see, the sidecar files are now perfectly supported...

**Comment by grasdk** [comment]
Yeah. Should be supported. Just discovered this ticket now. @bpatrik I guess this ticket can be closed.

**Comment by mcdamo** [comment]
Yeah great news, latest v2.0.3 reads from XMP files.

Video files are missing some details - caption, camera make, GPS location, ... but they do show the keyword tags. The missing details might be hidden by pigallery.

**Comment by Marxsal** [comment]
It seems to totally ignore XMP files for me. Using whatever latest docker pull is. I don't see any release notes claiming XMP is part of the gallery.  

Also seems to ignore title.

**Comment by grasdk** [comment]
@Marxsal 
Would you be able to provide an example. I guess the xmp file would be enough. I can pair it with a test image to see if I can reproduce your problem.

**Comment by Marxsal** [comment]
On second thought I went back and made sure to purge the meta data from its original image. But even after that, PG2 was still not picking up the XMP data. 

Note that when the info in this XMP is applied, the image should sort into 1960.
 
Also, as I mentioned before, there doesn't seem to be a release note about XMP. So I'm wondering if it was temporarily in the code but removed later when there were complications.

As you can see, the original image was a JPG, so probably a JPG should be used for testing.

Thanks!

[IMAG0001.JPG.xmp.zip](https://github.com/user-attachments/files/20312227/IMAG0001.JPG.xmp.zip)

**Comment by grasdk** [comment]
Hey. I can't reproduce your problem. This is the version og PiGallery2 I use:

![Image](https://github.com/user-attachments/assets/e8e0a616-ad49-4d21-8136-acd7c4802140)

Sceenshot of test image (from the test backend assets folder) without XMP-file:

![Image](https://github.com/user-attachments/assets/a3139eee-c82e-4958-a132-4868feae53f1)

Screenshot of same test image with xmp file:

![Image](https://github.com/user-attachments/assets/e01ab4a6-4d34-4a8e-9632-baa0af02b4cc)

I tried these two variants of xml-file naming:

`IMAG0001.jpg` paired with `IMAG0001.xmp`
`IMAG0001.JPG` paired with `IMAG0001.JPG.xmp`

boths variants had their metadata "overshadow" the embedded metadata.

**Comment by Marxsal** [comment]
How do you run the "edge" version?

**Comment by bpatrik** [comment]
If I understand correctly this feature fully landed
